packages/timeline-state-resolver/src/integrations/tcpSend/tcpConnection.ts

```diff
@@ -0,0 +1,154 @@
+import { EventEmitter } from 'eventemitter3'
+import { TCPSendOptions } from 'timeline-state-resolver-types'
+import { Socket } from 'net'
+
+const TIMEOUT = 3000 // ms
+const RETRY_TIMEOUT_FIRST = 500 // ms
+const RETRY_TIMEOUT_SUBSEQUENT = 5000 // ms
+export interface TcpConnectionEvents {
+	connectionChanged: [connected: boolean]
+	error: [context: string, error: Error]
+}
+
+export class TcpConnection extends EventEmitter<TcpConnectionEvents> {
+	/**
+	 * Is set when the connection is active.
+	 * is set to undefined if disconnect() has been called (then do not try to reconnect)
+	 */
+	private activeOptions: TCPSendOptions | undefined = undefined
+	private _tcpClient: Socket | null = null
+
+	private _connected = false
+	private _retryConnectTimeout: NodeJS.Timer
+
+	get connected(): boolean {
+		return this._connected
+	}
+	activate(options: TCPSendOptions): void {
+		this.activeOptions = options
+
+		this.ensureConnection().catch((err) => {
+			this.emit('error', 'activate TCP', err)
+		})
+	}
+	async ensureConnection(): Promise<Socket> {
+		if (!this.activeOptions) throw new Error('TCP connection not activated')
+		const activeOptions = this.activeOptions
+
+		return new Promise((resolve, reject) => {
+			if (!this._tcpClient) {
+				this._tcpClient = new Socket()
+				this._tcpClient.on('connect', () => {
+					this._setConnected(true)
+				})
+				this._tcpClient.on('close', () => {
+					this._setConnected(false)
+				})
+				this._tcpClient.on('end', () => {
+					this._setConnected(false)
+				})
+			}
+			const tcpClient: Socket = this._tcpClient
+
+			if (!this.connected) {
+				tcpClient.connect(activeOptions.port, activeOptions.host, () => {
+					resolve(tcpClient)
+				})
+				setTimeout(() => {
```

Comment:
should there also be a listener for `error` here, to reject if the connection fails with a better failure than the timeout?